core/predicate_check.go

```diff
@@ -18,50 +18,90 @@ import (
 
 var ErrMissingPredicateContext = errors.New("missing predicate context")
 
-// CheckPredicates verifies the predicates of [tx] and returns the result. Returning an error invalidates the block.
-func CheckPredicates(rules params.Rules, predicateContext *precompileconfig.PredicateContext, tx *types.Transaction) (predicate.PrecompileResults, error) {
-	// Check that the transaction can cover its IntrinsicGas (including the gas required by the predicate) before
-	// verifying the predicate.
-	intrinsicGas, err := IntrinsicGas(tx.Data(), tx.AccessList(), tx.To() == nil, rules)
+// CheckBlockPredicates verifies the predicates of a block of transactions and
+// returns the results.
+//
+// Returning an error invalidates the block.
+func CheckBlockPredicates(
+	rules params.Rules,
+	predicateContext *precompileconfig.PredicateContext,
+	txs []*types.Transaction,
+) (predicate.BlockResults, error) {
+	var results predicate.BlockResults
+	// TODO: Calculate tx predicates concurrently.
+	for _, tx := range txs {
+		txResults, err := CheckTxPredicates(rules, predicateContext, tx)
+		if err != nil {
+			return nil, err
+		}
+		results.Set(tx.Hash(), txResults)
+	}
+	return results, nil
+}
+
+// CheckTxPredicates verifies the predicates of a transaction and returns the
+// results.
+//
+// Returning an error invalidates the transaction.
+func CheckTxPredicates(
+	rules params.Rules,
+	predicateContext *precompileconfig.PredicateContext,
+	tx *types.Transaction,
+) (predicate.PrecompileResults, error) {
+	// Check that the transaction can cover its IntrinsicGas, including the gas
+	// required by the predicate, before verifying the predicate.
+	accessList := tx.AccessList()
+	intrinsicGas, err := IntrinsicGas(tx.Data(), accessList, tx.To() == nil, rules)
 	if err != nil {
 		return nil, err
 	}
-	if tx.Gas() < intrinsicGas {
-		return nil, fmt.Errorf("%w for predicate verification (%d) < intrinsic gas (%d)", ErrIntrinsicGas, tx.Gas(), intrinsicGas)
+	if gas := tx.Gas(); gas < intrinsicGas {
+		return nil, fmt.Errorf("%w for predicate verification (%d) < intrinsic gas (%d)",
+			ErrIntrinsicGas,
+			gas,
+			intrinsicGas,
+		)
 	}
```

Comment:
As a followup I'm going to look into if this check is actually needed. It seems dubious at best that this check is required, and it makes this function much more gross.